### PR TITLE
Feature/asv 1076 desing specs for comments in store

### PR DESCRIPTION
**Pull request by paquiquinho**
**What does this PR do?**

This PR aims to finish the comment system, both on ui and ux levels. It also builds upon the previous comments development with bugfixes and some flow changes.

Still missing behaviour for when someone tries to comment while logged out;

**Database changed?**

No

**Where should the reviewer start?**

- [ ] CommentFragment.java
- [ ] StoreLatestCommentsWidget.java

**How should this be manually tested?**

Try and use the comments on 3 levels: store, first level comment fragment (press the store comment widget more button), and second level comment fragment (replies). There is supposed to be only one level of depth in replies (reply only to main comment, no replies to replies). Also check the design matches zeplin;

**What are the relevant tickets?**

  Tickets related to this pull-request: [ASV-1076](<https://aptoide.atlassian.net/browse/ASV-1076>)

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Partners build
- [ ] Unit tests pass
- [ ] Functional QA tests pass

### CODE REVIEW

File: app/src/main/java/cm/aptoide/pt/view/recycler/displayable/DisplayablesFactory.java [hunk 1]
```diff
@@ -412,19 +418,18 @@ private static Displayable createRecommendedStores(GetStoreWidgets.WSWidget wsWi
...
-        .size() > 0) {
-      displayables.add(
-          new StoreLatestCommentsDisplayable(data.second.getId(), data.second.getName(),
-              comments.getDataList()
-                  .getList()));
-    } else {
-      displayables.add(new StoreAddCommentDisplayable(data.second.getId(), data.second.getName(),
-          StoreTheme.get(storeTheme)));
-    }
+    //if (comments != null
```
Comment (phillipsgoncalves) [inline app/src/main/java/cm/aptoide/pt/view/recycler/displayable/DisplayablesFactory.java]: comments

File: app/src/main/res/layout/add_comment_inner_item.xml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        android:layout_marginRight="8dp"
+        tools:src="@drawable/ic_user_circle"
+        />
+    <EditText
+        android:id="@+id/add_comment"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:background="@android:color/transparent"
+        android:gravity="center_vertical"
+        android:hint="Add a comment here..."
```
Comment (phillipsgoncalves) [inline app/src/main/res/layout/add_comment_inner_item.xml]: Not extracted string

File: app/src/test/java/cm/aptoide/pt/commentdetail/CommentDetailPresenterTest.java [hunk 3]
```diff
@@ -41,7 +44,8 @@
 
     CommentDetailViewModel viewModel =
         new CommentDetailViewModel("Filipe Gonçalves", "http://via.placeholder.com/350x150",
-            "Eu sou do Benfica", 7, new Date(), dataModelResponse.getReplies());
+            "Eu sou do Benfica", "http://via.placeholder.com/350x150", 7, new Date(),
```
Comment (phillipsgoncalves) [inline app/src/test/java/cm/aptoide/pt/commentdetail/CommentDetailPresenterTest.java]: This does not make any sense

**Comment by phillipsgoncalves** [comment]
Tests:
There is no shadow on the last comment of the store view

Logged in Flow:
-> If i post a comment and i try to navigate to my store, after posting the comment (pressing on the user of the "local comment") it will go to the app default error view (Is everyone trying to send apps at the same time? view)

-> If i go to the detail of a comment, should i still be able to navigate to the store of the user that posted the comment? At the moment if i press that user, it doesn't navigate anywhere.

-> Posted a comment on the More Comments view, tried to go to my store by clicking on my avatar from that comment and the app crashed.

Not Logged in flow:
   -> Crashes when i try to open a comment in the store view.

File: app/src/main/java/cm/aptoide/pt/comments/view/StoreLatestCommentsWidget.java [hunk 4]
```diff
@@ -1,175 +1,195 @@
...
+    if (view != null) {
+      InputMethodManager imm = (InputMethodManager) getActivityNavigator().getActivity()
+          .getSystemService(Context.INPUT_METHOD_SERVICE);
+      assert imm != null;
+      imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
     }
+  }
 
-    private Completable showSignInMessage(@NonNull final View view) {
-      // R.string.you_need_to_be_logged_in, R.string.login,
```
Comment (phillipsgoncalves) [inline app/src/main/java/cm/aptoide/pt/comments/view/StoreLatestCommentsWidget.java]: This is old code, but please remove it.. somebody forgot this

File: app/src/main/java/cm/aptoide/pt/ApplicationModule.java [hunk 5]
```diff
@@ -1454,10 +1459,11 @@ AnalyticsEventParametersNormalizer providesAnalyticsNormalizer() {
   @Singleton @Provides CommentsRepository providesCommentsRepository(@Named("pool-v7")
       BodyInterceptor<cm.aptoide.pt.dataprovider.ws.v7.BaseBody> bodyInterceptorPoolV7,
       @Named("default") OkHttpClient okHttpClient, Converter.Factory converterFactory,
-      TokenInvalidator tokenInvalidator, @Named("default") SharedPreferences sharedPreferences) {
+      TokenInvalidator tokenInvalidator, @Named("default") SharedPreferences sharedPreferences,
+      CommentMapper commentMapper) {
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/ApplicationModule.java]: fixed

File: app/src/main/java/cm/aptoide/pt/comment/CommentsFragment.java [hunk 6]
```diff
@@ -125,6 +135,16 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
     }
   }
 
+  @Override public void showCommentErrorSnack() {
+    hideKeyboard();
+    Snackbar.make(this.getView(), "error", BaseTransientBottomBar.LENGTH_SHORT);
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/comment/CommentsFragment.java]: fixed

File: app/src/main/java/cm/aptoide/pt/comment/CommentsFragment.java [hunk 7]
```diff
@@ -125,6 +135,16 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
     }
   }
 
+  @Override public void showCommentErrorSnack() {
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/comment/CommentsFragment.java]: fixed

File: app/src/main/java/cm/aptoide/pt/comment/CommentsFragment.java [hunk 8]
```diff
@@ -168,15 +188,20 @@ private boolean isEndReached() {
...
   }
 
+  @Override public void hideKeyboard() {
+    super.hideKeyboard();
+  }
+
   public void setupToolbar() {
 
-    toolbar.setTitle(getArguments().getString("storeName", ""));
+    toolbar.setTitle("Store comments");
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/comment/CommentsFragment.java]: fixed

File: app/src/main/java/cm/aptoide/pt/comment/CommentsPresenter.java [hunk 9]
```diff
@@ -39,8 +43,22 @@ public CommentsPresenter(CommentsView view, CommentsListManager commentsListMana
...
+                  if (account.isLoggedIn()) {
+                    return account;
+                  } else {
+                    return null;
+                  }
+                })
+                .filter(account -> account != null)
+                .observeOn(viewScheduler)
+                .flatMapCompletable(account -> {
+                  view.addLocalComment(comment, account);
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/comment/CommentsPresenter.java]: this should actually be done on a doOnCompleted() as we prolly don't want to show a local comment if problems happened posting the actual comment;

File: app/src/main/java/cm/aptoide/pt/comment/CommentsPresenter.java [hunk 10]
```diff
@@ -39,8 +43,22 @@ public CommentsPresenter(CommentsView view, CommentsListManager commentsListMana
   @VisibleForTesting public void postComment() {
     view.getLifecycleEvent()
         .filter(lifecycleEvent -> lifecycleEvent.equals(View.LifecycleEvent.CREATE))
-        .flatMap(created -> view.commentPost())
-        .flatMapCompletable(commentsListManager::postComment)
+        .flatMap(created -> view.commentPost()
+            .doOnNext(__ -> view.hideKeyboard())
+            .flatMap(comment -> accountManager.accountStatus()
+                .map(account -> {
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/comment/CommentsPresenter.java]: discussed

File: app/src/main/java/cm/aptoide/pt/comment/CommentsAdapter.java [hunk 11]
```diff
@@ -96,6 +106,11 @@ public void removeLoadMore() {
     }
   }
 
+  public void addSingleComment(Comment comment) {
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/comment/CommentsAdapter.java]: discussed

File: app/src/main/java/cm/aptoide/pt/commentdetail/CommentDetailPresenter.java [hunk 12]
```diff
@@ -37,6 +41,32 @@ public CommentDetailPresenter(CommentDetailView view, CommentDetailManager comme
         });
   }
 
+  @VisibleForTesting public void postComment() {
+    view.getLifecycleEvent()
+        .filter(lifecycleEvent -> lifecycleEvent.equals(View.LifecycleEvent.CREATE))
+        .flatMap(created -> view.commentClicked()
+            .doOnNext(__ -> view.hideKeyboard())
+            .flatMap(comment -> accountManager.accountStatus()
+                .map(account -> {
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/commentdetail/CommentDetailPresenter.java]: discussed

File: app/src/main/java/cm/aptoide/pt/commentdetail/CommentDetailFragment.java [hunk 13]
```diff
@@ -109,16 +143,42 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+        new User(-1, account.getAvatar(), account.getNickname()), 0, new Date()));
+  }
+
+  @Override public Observable<Comment> commentClicked() {
+    return postCommentClickEvent;
+  }
+
+  public void setupToolbar() {
+
+    toolbar.setTitle("Comment replies");
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/commentdetail/CommentDetailFragment.java]: fixed

File: app/src/main/java/cm/aptoide/pt/commentdetail/CommentDetailPresenter.java [hunk 14]
```diff
@@ -37,6 +41,32 @@ public CommentDetailPresenter(CommentDetailView view, CommentDetailManager comme
...
+                  if (account.isLoggedIn()) {
+                    return account;
+                  } else {
+                    return null;
+                  }
+                })
+                .filter(account -> account != null)
+                .observeOn(viewScheduler)
+                .flatMapCompletable(account -> {
+                  view.addLocalComment(comment, account);
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/commentdetail/CommentDetailPresenter.java]: same as before

File: app/src/main/java/cm/aptoide/pt/comments/view/StoreLatestCommentsWidget.java [hunk 15]
```diff
@@ -1,175 +1,195 @@
...
-                AccountAnalytics.AccountOrigins.LATEST_COMMENTS_STORE))
-            .show();
-      });
-    }
+  private void addLocalComment(Comment comment, Account account) {
+    commentsAdapter.addSingleComment(new Comment(comment.getId(), comment.getMessage(),
+        new User(-1, account.getAvatar(), account.getNickname()), 0, new Date()));
+  }
+
+  private void setAdapterWithComments(StoreLatestCommentsDisplayable displayable) {
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/comments/view/StoreLatestCommentsWidget.java]: fixed

File: app/src/main/java/cm/aptoide/pt/comments/view/StoreLatestCommentsWidget.java [hunk 16]
```diff
@@ -1,175 +1,195 @@
...
+    compositeSubscription.add(postComment.doOnNext(__ -> hideKeyboard())
+        .flatMap(comment -> accountManager.accountStatus()
+            .map(account -> {
+              if (account.isLoggedIn()) {
+                return account;
+              } else {
+                return null;
+              }
+            })
+            .filter(account -> account != null)
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/comments/view/StoreLatestCommentsWidget.java]: same as before

File: app/src/main/java/cm/aptoide/pt/comments/view/StoreLatestCommentsWidget.java [hunk 17]
```diff
@@ -1,175 +1,195 @@
...
-      return Completable.fromAction(() -> {
-        Snackbar.make(view, R.string.you_need_to_be_logged_in, Snackbar.LENGTH_LONG)
-            .setAction(R.string.login, snackView -> accountNavigator.navigateToAccountView(
-                AccountAnalytics.AccountOrigins.LATEST_COMMENTS_STORE))
-            .show();
-      });
-    }
+  private void addLocalComment(Comment comment, Account account) {
+    commentsAdapter.addSingleComment(new Comment(comment.getId(), comment.getMessage(),
+        new User(-1, account.getAvatar(), account.getNickname()), 0, new Date()));
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/comments/view/StoreLatestCommentsWidget.java]: fixed

File: app/src/main/java/cm/aptoide/pt/comments/view/StoreLatestCommentsWidget.java [hunk 18]
```diff
@@ -1,175 +1,195 @@
...
+        }, throwable -> {
+          throw new OnErrorNotImplementedException(throwable);
+        }));
+    recyclerView.setAdapter(commentsAdapter);
+  }
+
+  private void showEmptyState() {
+    int handsEmojiCode = 0x1F64C;
+    String handsEmoji = new String(Character.toChars(handsEmojiCode));
+    emptyState.setText("No comments in this store yet, be the first! " + handsEmoji);
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/comments/view/StoreLatestCommentsWidget.java]: fixed

File: app/src/main/java/cm/aptoide/pt/store/view/StoreFragment.java [hunk 19]
```diff
@@ -571,6 +567,10 @@ protected boolean displayHomeUpAsEnabled() {
     }
   }
 
+  protected boolean displayHomeUpFAsEnabled() {
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/store/view/StoreFragment.java]: fixed

File: app/src/main/java/cm/aptoide/pt/view/recycler/displayable/DisplayablesFactory.java (hunk 1 above)
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/view/recycler/displayable/DisplayablesFactory.java]: fixed

File: app/src/main/java/cm/aptoide/pt/comment/CommentsPresenter.java [hunk 20]
```diff
@@ -106,14 +123,27 @@ private void addComments(CommentsListViewModel model) {
...
+        .doOnNext(comment -> commentsNavigator.navigateToCommentView(comment,
+            commentsListManager.getStoreId()))
         .compose(view.bindUntilEvent(View.LifecycleEvent.DESTROY))
         .subscribe(commentId -> {
         }, throwable -> {
           throw new OnErrorNotImplementedException(throwable);
         });
   }
 
+  private void handleClickOnUser() {
```
Comment (phillipsgoncalves) [inline app/src/main/java/cm/aptoide/pt/comment/CommentsPresenter.java]: unused.

Codestyle above this line

**Pull request merged**